zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,194 @@
+//! Contains code that interfaces with the zcash_history crate from
+//! librustzcash.
+
+use std::{collections::HashMap, convert::TryInto, io};
+
+use crate::{
+    block::Block,
+    parameters::{ConsensusBranchId, Network},
+    sapling,
+    transaction::Transaction,
+};
+
+/// A MMR Tree using zcash_history::Tree.
+///
+/// Currently it should not be used as a long-term data structure because it
+/// may grow undefinitely.
+pub struct Tree {
+    network: Network,
+    tree: zcash_history::Tree,
+}
+
+/// An encoded tree Node.
+pub struct Node {
+    node: [u8; zcash_history::MAX_NODE_DATA_SIZE],
+}
+
+impl Node {
+    /// Convert a Node into a zcash_history::Entry.
+    fn to_entry(&self, branch_id: ConsensusBranchId) -> Result<zcash_history::Entry, io::Error> {
+        zcash_history::Entry::from_bytes(branch_id.into(), self.node)
+    }
+}
+
+impl Tree {
+    /// Create a MMR tree with the given length.
+    ///
+    /// The `peaks` are the peaks of the MMR tree to build and their position in the
+    /// array representation of the tree.
+    /// The `extra` are extra nodes that enable removing nodes from the tree, and their position.
+    ///
+    /// Note that the length is usually larger than the length of `peaks` and `extra`, since
+    /// you don't need to pass every node, just the peaks of the tree (plus extra).
+    fn new(
+        network: Network,
+        branch_id: ConsensusBranchId,
```

Comment:
Can we pass a `NetworkUpgrade` here, and use it to look up the consensus branch id?
That way, we can use the network upgrade from the calling code. (We usually have one around.)

It works something like:
```rust
network_upgrade.consensus_branch_id().expect("unexpected pre-Overwinter MMR history tree")
```